qcodes/instrument_drivers/Keysight/Keysight_81180A.py

```diff
@@ -500,9 +501,12 @@ def upload_waveform(self, waveform, segment_number=None):
         # Convert waveform to 12bit (min = 0, max = 2**12 - 1)
         if self.output_coupling.get_latest() == "DAC":
             # Voltage limits are +- 0.5V.
-            waveform_12bit = (2 ** 11 - 1) + (2 ** 12 - 1) * waveform
+            waveform_12bit = 2 ** 11 + (2 ** 12 - 1) * waveform
         else:
-            waveform_12bit = (2 ** 11 - 1) + (2 ** 10 - 1) * waveform
+            # The output waveform of the Keysight 81180A is limited by 2Vpp,
+            # hence max waveform amplitude is 1V. Also 0V corresponds to 2048, -1V -> 0, +1V -> 4095.
+            waveform_12bit = np.array([(2 ** 11 + (2 ** 11 - 1) * waveform_point) if (waveform_point >= 0)
+                                       else (2 ** 11 + 2 ** 11 * waveform_point) for waveform_point in waveform])
```

Comment:
I think this would be neater if we use some pre-built linear interpolation options.

```python
from scipy.interpolate import interp1d
self._interp = interp1d([-1, 1], [0, 0xFFF],)
self.voltage_to_bits = lambda x: self._interp(x).astype(np.int16)

self.voltage_to_bits(0.4)
```

You could also make the interpolation map and the `voltage_to_bits` a real function defined in the 81180A module, so you can easily use it in future for debugging.

Reply:
The original code in this PR is rather inefficient, it's doing a point-by-point comparison, creating a list, and finally converting that to an array.
I really like the solution of @maij. The code is clear, though it adds a bit of overhead (70 ms for 1M points on my pc), but that's acceptable.